File: libs/ml-kernel/src/handlers/predictions.rs
use std::collections::HashMap;

use axum::{
    Json,
    extract::{Path, State},
};
use chrono::Utc;
use serde_json::Value;
use sqlx::{FromRow, query_as};
use uuid::Uuid;

use crate::{
    AppState,
    domain::predictions,
    models::{
        deployment::TrafficSplitEntry,
        prediction::{
            BatchPredictionJob, CreateBatchPredictionRequest, ListBatchPredictionsResponse,
            RealtimePredictionRequest, RealtimePredictionResponse,
        },
    },
};

use super::{ServiceResult, bad_request, db_error, deserialize_json, not_found, to_json};

#[derive(Debug, FromRow)]
struct DeploymentRow {
    traffic_split: Value,
}

#[derive(Debug, FromRow)]
struct ModelVersionRow {
    id: Uuid,
    version_number: i32,
    schema: Value,
}

#[derive(Debug, FromRow)]
struct BatchPredictionRow {
    id: Uuid,
    deployment_id: Uuid,
    status: String,
    record_count: i64,
    output_destination: Option<String>,
    outputs: Value,
    created_at: chrono::DateTime<Utc>,
    completed_at: Option<chrono::DateTime<Utc>>,
}

fn to_batch_job(row: BatchPredictionRow) -> BatchPredictionJob {
    BatchPredictionJob {
        id: row.id,
        deployment_id: row.deployment_id,
        status: row.status,
        record_count: row.record_count,
        output_destination: row.output_destination,
        outputs: deserialize_json(row.outputs),
        created_at: row.created_at,
        completed_at: row.completed_at,
    }
}

async fn persist_realtime_inference(
    db: &sqlx::PgPool,
    deployment_id: Uuid,
    outputs: &[crate::models::prediction::PredictionOutput],
    predicted_at: chrono::DateTime<Utc>,
) -> Result<(), sqlx::Error> {
    sqlx::query(
        r#"
		INSERT INTO ml_batch_predictions (
			id,
			deployment_id,
			status,
			record_count,
			output_destination,
			outputs,
			created_at,
			completed_at
		)
		VALUES ($1, $2, 'realtime', $3, NULL, $4, $5, $6)
		"#,
    )
    .bind(Uuid::now_v7())
    .bind(deployment_id)
    .bind(outputs.len() as i64)
    .bind(to_json(outputs))
    .bind(predicted_at)
    .bind(Some(predicted_at))
    .execute(db)
    .await?;

    Ok(())
}

async fn load_deployment(
    db: &sqlx::PgPool,
    deployment_id: Uuid,
) -> Result<Option<DeploymentRow>, sqlx::Error> {
    query_as::<_, DeploymentRow>("SELECT traffic_split FROM ml_deployments WHERE id = $1")
        .bind(deployment_id)
        .fetch_optional(db)
        .await
}

async fn load_versions(
    db: &sqlx::PgPool,
    splits: &[TrafficSplitEntry],
) -> Result<HashMap<Uuid, predictions::ModelRuntime>, sqlx::Error> {
    let mut versions = HashMap::new();
    for split in splits {
        if versions.contains_key(&split.model_version_id) {
            continue;
        }

        if let Some(row) = query_as::<_, ModelVersionRow>(
            "SELECT id, version_number, schema FROM ml_model_versions WHERE id = $1",
        )
        .bind(split.model_version_id)
        .fetch_optional(db)
        .await?
        {
            versions.insert(
                row.id,
                predictions::ModelRuntime {
                    version_number: row.version_number,
                    schema: row.schema,
                },
            );
        }
    }

    Ok(versions)
}

pub async fn realtime_predict(
    State(state): State<AppState>,
    Path(deployment_id): Path<Uuid>,
    Json(body): Json<RealtimePredictionRequest>,
) -> ServiceResult<RealtimePredictionResponse> {
    if body.inputs.is_empty() {
        return Err(bad_request("prediction inputs are required"));
    }

    let Some(deployment) = load_deployment(&state.db, deployment_id)
        .await
        .map_err(|cause| db_error(&cause))?
    else {
        return Err(not_found("deployment not found"));
    };

    let splits: Vec<TrafficSplitEntry> = deserialize_json(deployment.traffic_split);
    if splits.is_empty() {
        return Err(bad_request("deployment has no traffic split configured"));
    }

    let version_map = load_versions(&state.db, &splits)
        .await
        .map_err(|cause| db_error(&cause))?;
    if version_map.is_empty() {
        return Err(not_found("deployment versions not found"));
    }

    let outputs = body
        .inputs
        .iter()
        .enumerate()
        .filter_map(|(index, input)| {
            let split = predictions::route_variant(&splits, index)?;
            let runtime = version_map.get(&split.model_version_id)?;
            Some(predictions::predict_record(
                input,
                &split,
                runtime,
                body.explain,
                index,
            ))
        })
        .collect::<Vec<_>>();

    let predicted_at = Utc::now();
    if let Err(error) =
        persist_realtime_inference(&state.db, deployment_id, &outputs, predicted_at).await
    {
        tracing::warn!(
            deployment_id = %deployment_id,
            "failed to persist realtime inference history: {error}"
        );
    }

    Ok(Json(RealtimePredictionResponse {
        deployment_id,
        outputs,
        predicted_at,
    }))
}

pub async fn list_batch_predictions(
    State(state): State<AppState>,
) -> ServiceResult<ListBatchPredictionsResponse> {
    let rows = query_as::<_, BatchPredictionRow>(
        r#"
		SELECT
			id,
			deployment_id,
			status,
			record_count,
			output_destination,
			outputs,
			created_at,
			completed_at
		FROM ml_batch_predictions
		ORDER BY created_at DESC
		"#,
    )
    .fetch_all(&state.db)
    .await
    .map_err(|cause| db_error(&cause))?;

    Ok(Json(ListBatchPredictionsResponse {
        data: rows.into_iter().map(to_batch_job).collect(),
    }))
}

pub async fn create_batch_prediction(
    State(state): State<AppState>,
    Json(body): Json<CreateBatchPredictionRequest>,
) -> ServiceResult<BatchPredictionJob> {
    if body.records.is_empty() {
        return Err(bad_request("batch prediction records are required"));
    }

    let Some(deployment) = load_deployment(&state.db, body.deployment_id)
        .await
        .map_err(|cause| db_error(&cause))?
    else {
        return Err(not_found("deployment not found"));
    };

    let splits: Vec<TrafficSplitEntry> = deserialize_json(deployment.traffic_split);
    if splits.is_empty() {
        return Err(bad_request("deployment has no traffic split configured"));
    }

    let version_map = load_versions(&state.db, &splits)
        .await
        .map_err(|cause| db_error(&cause))?;
    if version_map.is_empty() {
        return Err(not_found("deployment versions not found"));
    }

    let outputs = body
        .records
        .iter()
        .enumerate()
        .filter_map(|(index, input)| {
            let split = predictions::route_variant(&splits, index)?;
            let runtime = version_map.get(&split.model_version_id)?;
            Some(predictions::predict_record(
                input, &split, runtime, true, index,
            ))
        })
        .collect::<Vec<_>>();

    let row = query_as::<_, BatchPredictionRow>(
        r#"
		INSERT INTO ml_batch_predictions (
			id,
			deployment_id,
			status,
			record_count,
			output_destination,
			outputs,
			created_at,
			completed_at
		)
		VALUES ($1, $2, 'completed', $3, $4, $5, $6, $7)
		RETURNING
			id,
			deployment_id,
			status,
			record_count,
			output_destination,
			outputs,
			created_at,
			completed_at
		"#,
    )
    .bind(Uuid::now_v7())
    .bind(body.deployment_id)
    .bind(outputs.len() as i64)
    .bind(body.output_destination)
    .bind(to_json(&outputs))
    .bind(Utc::now())
    .bind(Some(Utc::now()))
    .fetch_one(&state.db)
    .await
    .map_err(|cause| db_error(&cause))?;

    Ok(Json(to_batch_job(row)))
}

File: openfoundry-go/libs/ml-kernel-go/handlers/predictions.go
package handlers

import (
	"context"
	"encoding/json"
	"errors"
	"net/http"
	"time"

	"github.com/google/uuid"
	"github.com/jackc/pgx/v5"
	"github.com/jackc/pgx/v5/pgxpool"

	"github.com/openfoundry/openfoundry-go/libs/ml-kernel-go/domain/predictions"
	"github.com/openfoundry/openfoundry-go/libs/ml-kernel-go/models"
)

// PredictionsHandlers exposes realtime + batch prediction endpoints.
// Mirrors libs/ml-kernel/src/handlers/predictions.rs:
//   - POST /api/v1/deployments/{id}/predictions    (realtime)
//   - GET  /api/v1/predictions/batch                (list)
//   - POST /api/v1/predictions/batch                (create)
type PredictionsHandlers struct {
	Pool *pgxpool.Pool
}

const batchPredictionColumns = `id, deployment_id, status, record_count,
                                output_destination, outputs,
                                created_at, completed_at`

type predictionsScanner interface {
	Scan(...any) error
}

func scanBatchJob(s predictionsScanner) (models.BatchPredictionJob, error) {
	var b models.BatchPredictionJob
	var outputDestination *string
	var outputsRaw []byte
	var completedAt *time.Time
	if err := s.Scan(
		&b.ID, &b.DeploymentID, &b.Status, &b.RecordCount,
		&outputDestination, &outputsRaw, &b.CreatedAt, &completedAt,
	); err != nil {
		return b, err
	}
	b.OutputDestination = outputDestination
	b.CompletedAt = completedAt
	if len(outputsRaw) > 0 {
		_ = json.Unmarshal(outputsRaw, &b.Outputs)
	}
	if b.Outputs == nil {
		b.Outputs = []models.PredictionOutput{}
	}
	return b, nil
}

func (h *PredictionsHandlers) loadDeploymentSplits(ctx context.Context, deploymentID uuid.UUID) ([]models.TrafficSplitEntry, error) {
	row := h.Pool.QueryRow(ctx,
		`SELECT traffic_split FROM ml_deployments WHERE id = $1`, deploymentID)
	var raw []byte
	if err := row.Scan(&raw); err != nil {
		return nil, err
	}
	var splits []models.TrafficSplitEntry
	if len(raw) > 0 {
		_ = json.Unmarshal(raw, &splits)
	}
	return splits, nil
}

func (h *PredictionsHandlers) loadVersions(ctx context.Context, splits []models.TrafficSplitEntry) (map[uuid.UUID]predictions.ModelRuntime, error) {
	versions := make(map[uuid.UUID]predictions.ModelRuntime, len(splits))
	for _, s := range splits {
		if _, ok := versions[s.ModelVersionID]; ok {
			continue
		}
		row := h.Pool.QueryRow(ctx,
			`SELECT id, version_number, schema FROM ml_model_versions WHERE id = $1`,
			s.ModelVersionID)
		var id uuid.UUID
		var versionNumber int32
		var schemaRaw []byte
		if err := row.Scan(&id, &versionNumber, &schemaRaw); err != nil {
			if errors.Is(err, pgx.ErrNoRows) {
				continue
			}
			return nil, err
		}
		var schema map[string]any
		if len(schemaRaw) > 0 {
			_ = json.Unmarshal(schemaRaw, &schema)
		}
		versions[id] = predictions.ModelRuntime{
			VersionNumber: versionNumber,
			Schema:        schema,
		}
	}
	return versions, nil
}

// RealtimePredict handles `POST /api/v1/deployments/{id}/predictions`.
func (h *PredictionsHandlers) RealtimePredict(w http.ResponseWriter, r *http.Request, deploymentID uuid.UUID) {
	var body models.RealtimePredictionRequest
	if err := json.NewDecoder(r.Body).Decode(&body); err != nil {
		writeError(w, http.StatusBadRequest, err.Error())
		return
	}
	if len(body.Inputs) == 0 {
		writeError(w, http.StatusBadRequest, "prediction inputs are required")
		return
	}

	splits, err := h.loadDeploymentSplits(r.Context(), deploymentID)
	if errors.Is(err, pgx.ErrNoRows) {
		writeError(w, http.StatusNotFound, "deployment not found")
		return
	}
	if err != nil {
		dbError(w, err)
		return
	}
	if len(splits) == 0 {
		writeError(w, http.StatusBadRequest, "deployment has no traffic split configured")
		return
	}

	versions, err := h.loadVersions(r.Context(), splits)
	if err != nil {
		dbError(w, err)
		return
	}
	if len(versions) == 0 {
		writeError(w, http.StatusNotFound, "deployment versions not found")
		return
	}

	outputs := runPredictions(body.Inputs, splits, versions, body.Explain)
	predictedAt := time.Now().UTC()

	if err := h.persistRealtimeInference(r.Context(), deploymentID, outputs, predictedAt); err != nil {
		// Match Rust: log + continue. The realtime path returns the
		// computed outputs even when the history-write fails.
		_ = err
	}

	writeJSON(w, http.StatusOK, models.RealtimePredictionResponse{
		DeploymentID: deploymentID,
		Outputs:      outputs,
		PredictedAt:  predictedAt,
	})
}

// ListBatchPredictions handles `GET /api/v1/predictions/batch`.
func (h *PredictionsHandlers) ListBatchPredictions(w http.ResponseWriter, r *http.Request) {
	rows, err := h.Pool.Query(r.Context(),
		`SELECT `+batchPredictionColumns+` FROM ml_batch_predictions
          ORDER BY created_at DESC`)
	if err != nil {
		dbError(w, err)
		return
	}
	defer rows.Close()
	out := make([]models.BatchPredictionJob, 0)
	for rows.Next() {
		j, err := scanBatchJob(rows)
		if err != nil {
			dbError(w, err)
			return
		}
		out = append(out, j)
	}
	writeJSON(w, http.StatusOK, models.ListBatchPredictionsResponse{Data: out})
}

// CreateBatchPrediction handles `POST /api/v1/predictions/batch`.
func (h *PredictionsHandlers) CreateBatchPrediction(w http.ResponseWriter, r *http.Request) {
	var body models.CreateBatchPredictionRequest
	if err := json.NewDecoder(r.Body).Decode(&body); err != nil {
		writeError(w, http.StatusBadRequest, err.Error())
		return
	}
	if len(body.Records) == 0 {
		writeError(w, http.StatusBadRequest, "batch prediction records are required")
		return
	}

	splits, err := h.loadDeploymentSplits(r.Context(), body.DeploymentID)
	if errors.Is(err, pgx.ErrNoRows) {
		writeError(w, http.StatusNotFound, "deployment not found")
		return
	}
	if err != nil {
		dbError(w, err)
		return
	}
	if len(splits) == 0 {
		writeError(w, http.StatusBadRequest, "deployment has no traffic split configured")
		return
	}

	versions, err := h.loadVersions(r.Context(), splits)
	if err != nil {
		dbError(w, err)
		return
	}
	if len(versions) == 0 {
		writeError(w, http.StatusNotFound, "deployment versions not found")
		return
	}

	outputs := runPredictions(body.Records, splits, versions, true)

	now := time.Now().UTC()
	outputsJSON, _ := json.Marshal(outputs)
	row := h.Pool.QueryRow(r.Context(),
		`INSERT INTO ml_batch_predictions
              (id, deployment_id, status, record_count,
               output_destination, outputs, created_at, completed_at)
            VALUES ($1, $2, 'completed', $3, $4, $5, $6, $7)
            RETURNING `+batchPredictionColumns,
		uuid.New(), body.DeploymentID, int64(len(outputs)),
		body.OutputDestination, outputsJSON, now, now)
	job, err := scanBatchJob(row)
	if err != nil {
		dbError(w, err)
		return
	}
	writeJSON(w, http.StatusOK, job)
}

func (h *PredictionsHandlers) persistRealtimeInference(ctx context.Context, deploymentID uuid.UUID, outputs []models.PredictionOutput, predictedAt time.Time) error {
	outputsJSON, _ := json.Marshal(outputs)
	_, err := h.Pool.Exec(ctx,
		`INSERT INTO ml_batch_predictions
              (id, deployment_id, status, record_count,
               output_destination, outputs, created_at, completed_at)
            VALUES ($1, $2, 'realtime', $3, NULL, $4, $5, $6)`,
		uuid.New(), deploymentID, int64(len(outputs)),
		outputsJSON, predictedAt, predictedAt)
	return err
}

// runPredictions applies route_variant + predict_record to each input
// record. Mirrors the filter_map iteration in the Rust source —
// records whose split or runtime can't be resolved are silently
// dropped, matching wire-compat semantics.
func runPredictions(inputs []json.RawMessage, splits []models.TrafficSplitEntry, versions map[uuid.UUID]predictions.ModelRuntime, explain bool) []models.PredictionOutput {
	outputs := make([]models.PredictionOutput, 0, len(inputs))
	for index, raw := range inputs {
		split, ok := predictions.RouteVariant(splits, index)
		if !ok {
			continue
		}
		runtime, ok := versions[split.ModelVersionID]
		if !ok {
			continue
		}
		var input any
		if len(raw) > 0 {
			_ = json.Unmarshal(raw, &input)
		}
		outputs = append(outputs, predictions.PredictRecord(input, split, runtime, explain, index))
	}
	return outputs
}
